Microsoft.Toolkit.Uwp/IncrementalLoadingCollection/IncrementalLoadingCollection.cs

```diff
@@ -0,0 +1,162 @@
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH
+// THE CODE OR THE USE OR OTHER DEALINGS IN THE CODE.
+// ******************************************************************
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Linq;
+using System.Runtime.InteropServices.WindowsRuntime;
+using System.Threading;
+using System.Threading.Tasks;
+using Windows.Foundation;
+using Windows.UI.Core;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Data;
+
+namespace Microsoft.Toolkit.Uwp
+{
+    /// <summary>
+    /// This class represents an <see cref="ObservableCollection{IType}"/> whose items can be loaded incrementally.
+    /// </summary>
+    /// <typeparam name="TSource">
+    /// The data source that must be loaded incrementally.
+    /// </typeparam>
+    /// <typeparam name="IType">
+    /// The type of collection items.
+    /// </typeparam>
+    /// <seealso cref="IIncrementalSource{TSource}"/>
+    /// <seealso cref="ISupportIncrementalLoading"/>
+    public class IncrementalLoadingCollection<TSource, IType> : ObservableCollection<IType>,
+         ISupportIncrementalLoading
+         where TSource : IIncrementalSource<IType>, new()
+    {
+        private readonly TSource _source;
+
+        private readonly int _itemsPerPage;
+        private int _currentPageIndex;
+
+        private CancellationToken _cancellationToken;
+
+        private bool _isLoading;
+
+        /// <summary>
+        /// Gets a value indicating whether new items are being loaded.
+        /// </summary>
+        public bool IsLoading
+        {
+            get
+            {
+                return _isLoading;
+            }
+
+            private set
+            {
+                if (value != _isLoading)
+                {
+                    _isLoading = value;
+                    OnPropertyChanged(new PropertyChangedEventArgs(nameof(IsLoading)));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="IncrementalLoadingCollection{TSource, IType}"/> class, optionally specifying how many items to load for each data page.
+        /// </summary>
+        /// <param name="itemsPerPage">
+        /// The number of items to retrieve for each call. Default is 20.
+        /// </param>
+        public IncrementalLoadingCollection(int itemsPerPage = 20)
+        {
+            _source = new TSource();
+            _itemsPerPage = itemsPerPage;
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the collection contains more items to retrieve.
+        /// </summary>
+        public bool HasMoreItems
+        {
+            get
+            {
+                if (_cancellationToken.IsCancellationRequested)
```

Comment:
If _cancellationToken is null (which it currently would be by default), this would crash the app.